Dashboard item description (1/2)

liyiy opened the pull request:
## Changes

(1/2) Updates for dashboard item editing on #3551  

<img width="1273" alt="Screen Shot 2021-05-06 at 3 06 17 PM" src="https://user-images.githubusercontent.com/25164963/117352161-a574b900-ae7c-11eb-8de0-df66ea3f7397.png">

- Follow up PR (2/2) will include 
     - allowing dashboard item title to be edited 
     - some more style fixes
     - e2es

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious

liyiy committed:
enable dashboard item descriptions for premium

liyiy committed:
work out kea logic for dashboard item edit mode

liyiy committed:
set up frontend for dashboard item editing

liyiy committed:
update item description view and edit button

liyiy committed:
dashboard item scss file

liyiy committed:
style updates and render description correctly after saving changes

liyiy committed:
allow editing for premium only

liyiy committed:
update placeholder text

liyiy commented:
There's a lot of the same code between the dashboard description edit and dashboard item description edit, maybe worth going over at a later time to see which parts to consolidate ^_^

liyiy commented:
confused with prettier failure that doesn't show what the error is in the `Dashboard.scss` file 😦

paolodamico committed:
fix leftover test condition

paolodamico committed:
Merge branch 'master' into dashboard-item-description

paolodamico commented on frontend/src/models/dashboardItemsModel.tsx:
```diff
@@ -3,23 +3,34 @@ import { kea } from 'kea'
     actions: () => ({
         renameDashboardItem: (item: DashboardItemType) => ({ item }),
         renameDashboardItemSuccess: (item: DashboardItemType) => ({ item }),
+        updateDashboardItem: (id: number, payload: any) => ({ id, ...payload }),
```
The payload could be a `Partial<DashboardItemType>` probably

liyiy replied:
I'm getting `✖ flake8 --select=E9,F63,F7,F82 failed without output (ENOENT)` . errors when I try to merge this branch with master, how do I resolve this?

paolodamico replied:
Hmmm not sure what's going on there, but I fix the prettier error and merged master. Sometimes when you try to commit merge commits, husky will fail for some reason, you can try finishing the commit with `--no-verify`.

paolodamico reviewed (changes requested):
Looking pretty good!! Reviewed this from a UX perspective,

1. I think we should keep the same consistent behavior for editing as we do for dashboards (see below). Clicking to edit instead of the button, close if you click outside, cmd/cntrl + enter to submit.
2. I think dashboard item might be confusing. Maybe dashboard insight?
3. Related to the above the placeholder text should be: "Add a description for this dashboard insight..."
4. The description should not be muted if it's set (show in default text color). Probably worth having it the same font size as the rest (unsure, let's try it out).
5. You should be able to edit the item's description from the dashboard page.

macobo commented on frontend/src/models/dashboardItemsModel.tsx:
```diff
@@ -3,23 +3,34 @@ import { kea } from 'kea'
     }),
+    reducers: () => ({
+        dashboardItemMode: [
+            null,
```
You probably want to set an explicit type here: `null as DashboardItemMode | null`. In general we should avoid implicit any's in typescript.

macobo commented on frontend/src/models/dashboardItemsModel.tsx:
```diff
@@ -34,6 +45,15 @@ export const dashboardItemsModel = kea<dashboardItemsModelType<DashboardItemType
+            const response = await api.update(`api/dashboard_item/${id}`, payload)
+            actions.setDashboardItemMode(null)
+            insightLogic.actions.setDashboardItem(response)
+            return response
```
Q: is the return needed here? Why?

liyiy replied:
nope! Thought I would need it but I didn't 😃 I shall remove it

liyiy committed:
rename dashboard item to dashboard insight

liyiy committed:
typing and naming updates

liyiy committed:
click to edit item description in box

liyiy committed:
clean up

liyiy committed:
remove unnecessary return

paolodamico committed:
fix prettier

paolodamico committed:
Merge branch 'master' into dashboard-item-description

liyiy committed:
Merge changes into branch

liyiy committed:
streamline description edit UX

liyiy committed:
create userLogic selector for repeated user dashboard collab feature check

liyiy committed:
Merge branch 'master' into dashboard-item-description

liyiy commented:
@macobo I addressed most of the concerns on @paolodamico's comment but in favor of keeping this PR from spiraling, I will address everything else in the second PR of this (editing dashboard insight description on the dashboards page)
